mm2src/mm2_event_stream/src/manager.rs

```diff
@@ -0,0 +1,515 @@
+    /// Couldn't spawn the streamer.
+    SpawnError(String),
+    /// The client is not known/registered.
+    UnknownClient,
+    /// A client with the same ID already exists.
+    ClientExists,
+    /// The client is already listening to the streamer.
+    ClientAlreadyListening,
+}
+
+struct StreamerInfo {
+    /// The communication channel to the streamer.
+    data_in: Option<UnboundedSender<Box<dyn Any + Send>>>,
+    /// Clients the streamer is serving for.
+    clients: HashSet<u64>,
+    /// The shutdown handle of the streamer.
+    shutdown: oneshot::Sender<()>,
+}
+
+impl StreamerInfo {
+    fn new(data_in: Option<UnboundedSender<Box<dyn Any + Send>>>, shutdown: oneshot::Sender<()>) -> Self {
+        Self {
+            data_in,
+            clients: HashSet::new(),
+            shutdown,
+        }
+    }
+
+    fn add_client(&mut self, client_id: u64) { self.clients.insert(client_id); }
+
+    fn remove_client(&mut self, client_id: &u64) { self.clients.remove(client_id); }
+
+    fn is_down(&self) -> bool { self.shutdown.is_canceled() }
+}
+
+struct ClientInfo {
+    /// The streamers the client is listening to.
+    listening_to: HashSet<String>,
+    /// The communication/stream-out channel to the client.
+    // NOTE: Here we are using `tokio`'s `mpsc` because the one in `futures` have some extra feature
+    // (ref: https://users.rust-lang.org/t/why-does-try-send-from-crate-futures-require-mut-self/100389).
+    // This feature is aimed towards the multi-producer case (which we don't use) and requires a mutable
+    // reference on `try_send` calls. This will require us to put the channel in a mutex and degrade the
+    // broadcasting performance.
+    channel: mpsc::Sender<Arc<Event>>,
+}
+
+impl ClientInfo {
+    fn new(channel: mpsc::Sender<Arc<Event>>) -> Self {
+        Self {
+            listening_to: HashSet::new(),
+            channel,
+        }
+    }
+
+    fn add_streamer(&mut self, streamer_id: String) { self.listening_to.insert(streamer_id); }
+
+    fn remove_streamer(&mut self, streamer_id: &str) { self.listening_to.remove(streamer_id); }
+
+    fn listens_to(&self, streamer_id: &str) -> bool { self.listening_to.contains(streamer_id) }
+
+    fn send_event(&self, event: Arc<Event>) {
+        // Only `try_send` here. If the channel is full (client is slow), the message
+        // will be dropped and the client won't receive it.
+        // This avoids blocking the broadcast to other receivers.
+        self.channel.try_send(event).ok();
+    }
+}
+
+#[derive(Default)]
+struct StreamingManagerInner {
+    /// A map from streamer IDs to their communication channels (if present) and shutdown handles.
+    streamers: HashMap<String, StreamerInfo>,
+    /// An inverse map from client IDs to the streamers they are listening to and the communication channel with the client.
+    clients: HashMap<u64, ClientInfo>,
+}
+
+#[derive(Clone, Default)]
+pub struct StreamingManager(Arc<RwLock<StreamingManagerInner>>);
```

Comment:
I did some reading about wasm running in browser and I understand that it is generally not okay to use sync locks there (https://rustwasm.github.io/docs/wasm-bindgen/examples/raytrace.html#caveats).
Maybe better to use tokio async locks for wasm case?

Reply:
Looks like wasm modules currently are executed in same thread so locks in wasm do not play any role